src/common/src/util/value_encoding/row_encoding.rs

```diff
@@ -0,0 +1,214 @@
+        assert!(
+            self.buf.is_empty(),
+            "should not encode one RowEncoding object multiple times."
+        );
+        let mut offset_usize = vec![];
+        for datum in datum_refs {
+            offset_usize.push(self.buf.len());
+            if let Some(v) = datum.to_datum_ref() {
+                serialize_scalar(v, &mut self.buf);
+            }
+        }
+        let max_offset = *offset_usize
+            .last()
+            .expect("should encode at least one column");
+        self.set_big(&offset_usize, max_offset);
+    }
+}
+
+/// Column-Aware `Serializer` holds schema related information, and shall be
+/// created again once the schema changes
+pub struct Serializer {
+    encoded_column_ids: Vec<u8>,
+    datum_num: usize,
+    encoded_datum_num: Vec<u8>,
+}
+
+impl Serializer {
+    pub fn new(column_ids: &[ColumnId]) -> Self {
+        // currently we hard-code ColumnId as i32
+        let mut encoded_column_ids = Vec::with_capacity(column_ids.len() * 4);
+        for id in column_ids {
+            encoded_column_ids.put_i32_le(id.get_id());
+        }
+        let datum_num = column_ids.len();
+        let mut encoded_datum_num = vec![];
+        encoded_datum_num.put_u32_le(datum_num as u32);
+        Self {
+            encoded_column_ids,
+            datum_num,
+            encoded_datum_num,
+        }
+    }
+
+    pub fn serialize_row_column_aware(&self, row: impl Row) -> Vec<u8> {
+        assert_eq!(row.len(), self.datum_num);
+        let mut encoding = RowEncoding::new();
+        encoding.encode(row.iter());
+        self.serialize(encoding)
+    }
+
+    fn serialize(&self, encoding: RowEncoding) -> Vec<u8> {
+        let mut row_bytes = vec![];
+        row_bytes.put_u8(encoding.flag);
+        row_bytes.extend(self.encoded_datum_num.iter());
+        row_bytes.extend(self.encoded_column_ids.iter());
+        row_bytes.extend(encoding.offsets.iter());
+        row_bytes.extend(encoding.buf.iter());
+
+        row_bytes
+    }
+}
+
+/// Column-Aware `Deserializer` holds needed `ColumnIds` and their corresponding schema
+/// Should non-null default values be specified, a new field could be added to Deserializer
+pub struct Deserializer<'a> {
+    needed_column_ids: BTreeMap<i32, usize>,
+    schema: &'a [DataType],
+}
+
+impl<'a> Deserializer<'a> {
+    pub fn new(column_ids: &'a [ColumnId], schema: &'a [DataType]) -> Self {
+        assert_eq!(column_ids.len(), schema.len());
+        Self {
+            needed_column_ids: column_ids
+                .iter()
+                .enumerate()
+                .map(|(i, c)| (c.get_id(), i))
+                .collect::<BTreeMap<_, _>>(),
+            schema,
+        }
+    }
+
+    pub fn decode(&self, mut encoded_bytes: &[u8]) -> Result<Vec<Datum>> {
+        let flag = encoded_bytes.get_u8();
+        let offset_bytes = match flag & 0b11 {
+            0b01 => 1,
+            0b10 => 2,
+            0b11 => 4,
```

Comment:
IMHO it's not a good idea to use binary literals here and in `set_big`.

Reply:
Will try to use `bitflags` 😄